hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/client/OMBlockPrefetchClient.java

```diff
@@ -0,0 +1,369 @@
+          ExpiringAllocatedBlock expiringBlock = queue.poll();
+          if (expiringBlock == null) {
+            break;
+          }
+
+          if (now > expiringBlock.getExpiryTime()) {
+            continue;
+          }
+
+          tempValidBlocks.add(expiringBlock);
+        }
+
+        // If there aren't enough blocks in cache, we fallback to SCM.
+        if (tempValidBlocks.size() < numBlocks) {
+          List<AllocatedBlock> newBlocks = scmBlockLocationProtocol.allocateBlock(scmBlockSize, numBlocks,
+              replicationConfig, serviceID, excludeList, clientMachine);
+          allocatedBlocks.addAll(newBlocks);
+
+          // Return unused valid blocks back to the front of the queue (preserving original order).
+          for (int i = tempValidBlocks.size() - 1; i >= 0; i--) {
+            queue.addFirst(tempValidBlocks.get(i));
+          }
+          metrics.incrementCacheMisses();
+        } else {
+          for (ExpiringAllocatedBlock expiringBlock : tempValidBlocks) {
+            AllocatedBlock block = expiringBlock.getBlock();
+            List<DatanodeDetails> sortedNodes =
+                sortDatanodes(block.getPipeline().getNodes(), clientMachine, clusterMap);
+            if (!Objects.equals(sortedNodes, block.getPipeline().getNodesInOrder())) {
+              block = block.toBuilder()
+                  .setPipeline(block.getPipeline().copyWithNodesInOrder(sortedNodes))
+                  .build();
+            }
+            allocatedBlocks.add(block);
+          }
+          metrics.incrementCacheHits();
+        }
+
+        int queueSize = queue.size();
+        if (queueSize < minBlocks) {
+          int blocksToPrefetch = maxBlocks - queueSize;
+          LOG.debug(
+              "Cache for {} is below threshold (size: {}, min: {}). Submitting async prefetch task for {} blocks.",
+              replicationConfig, queueSize, minBlocks, blocksToPrefetch);
+          submitPrefetchTask(scmBlockSize, blocksToPrefetch, replicationConfig, serviceID);
+        }
+
+        metrics.addReadFromQueueLatency(Time.monotonicNowNanos() - readStartTime);
+        return allocatedBlocks;
+
+      } else {
+        LOG.debug("Bypassing cache for {}. Reason: {}", replicationConfig, queue == null ?
+            "Unsupported replication config for caching." : "ExcludeList provided.");
+        metrics.addReadFromQueueLatency(Time.monotonicNowNanos() - readStartTime);
+        metrics.incrementCacheMisses();
+        return scmBlockLocationProtocol.allocateBlock(scmBlockSize, numBlocks, replicationConfig, serviceID,
+            excludeList, clientMachine);
+      }
+    } else {
+      return scmBlockLocationProtocol.allocateBlock(scmBlockSize, numBlocks, replicationConfig, serviceID, excludeList,
+          clientMachine);
+    }
+  }
+
+  private void submitPrefetchTask(long blockSize, int blocksToPrefetch, ReplicationConfig repConfig, String serviceID) {
+
+    if (!isPrefetching.compareAndSet(false, true)) {
```

Comment:
We need to account for replication configuration for pre fetching. Not all pre fetching tasks are the same.

Reply:
As per an earlier discussion, the idea was to regulate the overall load on SCM during prefetching (across different replication configurations).